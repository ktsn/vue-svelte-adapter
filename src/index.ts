import Vue, { VueConstructor, PropOptions } from 'vue'

export type Prop<T> = { (): T } | { new (...args: any[]): T & object }

export type PropValidator<T> = PropOptions<T> | Prop<T> | Prop<T>[]

export type RecordPropsDefinition<T> = { [K in keyof T]: PropValidator<T[K]> }

export type ArrayPropsDefinition<T> = (keyof T)[]

export type PropsDefinition<T> =
  | ArrayPropsDefinition<T>
  | RecordPropsDefinition<T>

export interface AdapterOptions<Props> {
  props?: PropsDefinition<Props>
}

export function toVue<Props>(
  Ctor: any,
  options: AdapterOptions<Props> = {}
): VueConstructor<Props & Vue> {
  return Vue.extend<Vue, {}, {}, Props>({
    props: options.props as any,

    mounted() {
      const instance = ((this as any).instance = new Ctor({
        target: this.$el,
        data: this.$props
      }))

      patchSvelteInstance(this, instance)

      observePropsDiff(this, diff => {
        instance.set(diff)
      })
    },

    beforeDestroy() {
      const { instance } = this as any
      instance.destroy()
    },

    render(h) {
      return h('div')
    }
  })
}

function noop(): void {
  // do nothing
}

function observePropsDiff(
  vm: Vue,
  cb: (diff: Record<string, any>) => void
): void {
  const props = vm.$props || {}
  const propKeys = Object.keys(props)
  let prevProps = copy(props)

  if (propKeys.length === 0) return

  vm.$watch(() => {
    const diff: Record<string, any> = {}

    propKeys.forEach(key => {
      const value = vm.$props[key]
      if (
        prevProps[key] !== value ||
        (value !== null && typeof value === 'object')
      ) {
        diff[key] = value
      }
    })

    if (Object.keys(diff).length > 0) {
      cb(diff)
      prevProps = copy(vm.$props || {})
    }
  }, noop)
}

function patchSvelteInstance(vm: Vue, svelte: any): void {
  const originalFire = svelte.fire
  svelte.fire = (eventName: string, data: any): void => {
    vm.$emit(eventName, data)
    return originalFire.call(svelte, eventName, data)
  }
}

function copy<T extends object>(value: T): T {
  return Object.keys(value).reduce<any>((acc, key) => {
    acc[key] = (value as any)[key]
    return acc
  }, {})
}
