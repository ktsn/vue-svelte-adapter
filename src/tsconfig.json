{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "declaration": true
  },
  "include": [
    "**/*.ts"
  ]
}