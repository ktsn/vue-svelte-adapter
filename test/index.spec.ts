import Vue, { VNode } from 'vue'
import { mount } from '@vue/test-utils'
import { toVue } from '../src/index'
import Basic from './fixtures/Basic.html'
import Data from './fixtures/Data.html'
import ObjectData from './fixtures/ObjectData.html'
import Event from './fixtures/Event.html'
import CheckDestroy from './fixtures/CheckDestroy.html'

describe('Vue Svelte Adapter', () => {
  it('renders template', () => {
    const BasicVue = toVue(Basic)
    const wrapper = mount(BasicVue)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('ports Vue props to Svelte data', () => {
    const DataVue = toVue(Data, {
      props: {
        message: String
      }
    })
    const wrapper = mount(DataVue, {
      propsData: {
        message: 'Test Message'
      }
    })
    expect(wrapper.text()).toBe('Message: Test Message')
  })

  it('updates Svelte data when props are updated', () => {
    const DataVue = toVue(Data, {
      props: {
        message: String
      }
    })
    const wrapper = mount(DataVue, {
      propsData: {
        message: 'Test'
      }
    })
    wrapper.setProps({
      message: 'Updated'
    })
    expect(wrapper.text()).toBe('Message: Updated')
  })

  it('updates Svelte data when object props are mutated', () => {
    const DataVue = toVue(ObjectData, {
      props: {
        user: Object
      }
    })
    const Wrapper = Vue.extend({
      data() {
        return {
          user: {
            name: 'Foo'
          }
        }
      },

      render(h): VNode {
        return h(DataVue, {
          props: {
            user: this.user
          }
        })
      }
    })
    const wrapper = mount(Wrapper)

    expect(wrapper.text()).toBe('Name: Foo')
    wrapper.vm.user.name = 'Bar'
    expect(wrapper.text()).toBe('Name: Bar')
  })

  it('observes component events', () => {
    const EventVue = toVue(Event)
    const wrapper = mount(EventVue)
    wrapper.find('button').trigger('click')
    expect(wrapper.emitted('test')[0]).toEqual(['clicked'])
  })

  it('teardowns Svelte component', () => {
    const spy = jest.fn()
    CheckDestroy.registerSpy(spy)

    const CheckDestroyVue = toVue(CheckDestroy)
    const wrapper = mount(CheckDestroyVue)

    expect(spy).not.toHaveBeenCalled()
    wrapper.destroy()
    expect(spy).toHaveBeenCalled()
  })
})
