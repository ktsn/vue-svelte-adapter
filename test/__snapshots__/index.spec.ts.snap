// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vue Svelte Adapter does not detach DOM nodes on destroy 1`] = `"<div><h1>Hello</h1></div>"`;

exports[`Vue Svelte Adapter renders template 1`] = `"<div><h1>Hello</h1></div>"`;
